### PR TITLE
Fix serialization of IndexFormatTooNewException and IndexFormatTooOldException

### PR DESCRIPTION
**Pull request by imotov**
This is essentially an ugly hack to get us by until a proper solution is possible with Lucene 5.3.

The current serialization of IndexFormatTooNewException and IndexFormatTooOldException exceptions is causing them to change the message every time they get serialized. So, while exception starts on the data node as 

`Format version is not supported (resource Format version is not supported (resource BufferedChecksumIndexInput(MockIndexInputWrapper(NIOFSIndexInput(path=\"/home/jenkins/workspace/es_core_master_metal/core/target/J2/temp/org.elasticsearch.index.store.CorruptedFileTest_40DD3FE2206C616D-001/tempDir-001/data/SUITE-slave19.build.ci.hetz.es.io-CHILD_VM=[2]-CLUSTER_SEED=[-5959755590901272649]-HASH=[6F0DDBF15A26B2]-cluster/nodes/0/indices/test/0/index/_0.si\")))): 257 (needs to be between 0 and 1)`

when it reaches the master and master publishes it to all the nodes it becomes:

`Format version is not supported (resource Format version is not supported (resource Format version is not supported (resource Format version is not supported (resource BufferedChecksumIndexInput(MockIndexInputWrapper(NIOFSIndexInput(path=\"/home/jenkins/workspace/es_core_master_metal/core/target/J2/temp/org.elasticsearch.index.store.CorruptedFileTest_40DD3FE2206C616D-001/tempDir-001/data/SUITE-slave19.build.ci.hetz.es.io-CHILD_VM=[2]-CLUSTER_SEED=[-5959755590901272649]-HASH=[6F0DDBF15A26B2]-cluster/nodes/0/indices/test/0/index/_0.si\")))): 257 (needs to be between 0 and 1)): -1 (needs to be between -1 and -1)): -1 (needs to be between -1 and -1)): -1 (needs to be between -1 and -1)];`

Basically on each iteration it gets wrapped into `Format version is not supported (` and `): -1 (needs to be between -1 and -1)`. When this exception becomes a part of a cluster state it trips the cluster state consistency check since data nodes have one extra  `Format version is not supported (` comparing to the data node. An example of such failure can be found in http://build-us-00.elastic.co/job/es_core_master_metal/10331/ .

### CODE REVIEW

File: core/src/main/java/org/elasticsearch/common/io/stream/StreamOutput.java [hunk 1]
```diff
@@ -453,19 +456,85 @@ public void writeOptionalStreamable(@Nullable Streamable streamable) throws IOEx
         }
     }
 
+    // Remove after upgrade to lucene 5.3
```
Comment (dakrone) [inline core/src/main/java/org/elasticsearch/common/io/stream/StreamOutput.java]: Can you add a static assertion that will fail when we upgrade to Lucene 5.3 like we do for X-classes?

**Comment by dakrone** [comment]
Can you also attach a link to the build failure this cause(s)?

**Comment by dakrone** [comment]
I also think this needs more documentation so after the upgrade someone can undo it, as-is, there's no explanation for what exactly is happening (multiple wrapping of the exception)

**Comment by imotov** [comment]
It causes http://build-us-00.elastic.co/job/es_core_master_metal/10331/ to fail. Hmm, I thought "Lucene 5.3 will have getters for all these" is explaining enough, I guess not. Will try to come up with more elaborate explanation.

**Comment by imotov** [comment]
@dakrone I added more detailed explanation of what needs to be done when we upgrade to lucene 5.3, added static check for the lucene version and updated PR comment to include description of the build failure this issue was causing. Could you take another look? Thanks!

@s1monw I trusted your comment in the original code that lucene 5.3 will get getters, but I cannot find the LUCENE ticket I could attach to the issue. Do you know where this work is getting tracked?

**Comment by dakrone** [comment]
LGTM

File: core/src/main/java/org/elasticsearch/common/io/stream/StreamOutput.java [hunk 2]
```diff
@@ -453,19 +456,100 @@ public void writeOptionalStreamable(@Nullable Streamable streamable) throws IOEx
...
     public void writeThrowable(Throwable throwable) throws IOException {
         if (throwable == null) {
             writeBoolean(false);
         } else {
             writeBoolean(true);
             boolean writeCause = true;
+            boolean writeMessage = true;
             if (throwable instanceof CorruptIndexException) {
                 writeVInt(1);
+                // Lucene 5.3 will have getters for all these
```
Comment (jpountz) [inline core/src/main/java/org/elasticsearch/common/io/stream/StreamOutput.java]: When I read this comment, I thought it was already committed to Lucene but I can't see it?

Comment (imotov) [inline core/src/main/java/org/elasticsearch/common/io/stream/StreamOutput.java]: @jpountz I copied this statement from @s1monw's code. Trying to find some evidence - please see my last comment above :)
